src/main/webapp/resources/js/pages/projects/samples-metadata-import/services/importReducer.ts

```diff
@@ -0,0 +1,251 @@
+import {
+  createAction,
+  createAsyncThunk,
+  createReducer,
+} from "@reduxjs/toolkit";
+import { validateSampleName } from "../../../../apis/metadata/sample-utils";
+import {
+  createSample,
+  FieldUpdate,
+  MetadataItem,
+  updateSample,
+  ValidateSampleNameModel,
+  validateSamples,
+  ValidateSamplesResponse,
+} from "../../../../apis/projects/samples";
+import { ImportDispatch, ImportState } from "../store";
+
+interface MetadataValidateDetailsItem {
+  isSampleNameValid: boolean;
+  foundSampleId?: number;
+}
+
+interface MetadataSaveDetailsItem {
+  saved: boolean;
+  error?: string;
+}
+interface SaveMetadataResponse {
+  metadataSaveDetails: Record<string, MetadataSaveDetailsItem>;
+}
+
+interface SetSampleNameColumnResponse {
+  sampleNameColumn: string;
+  metadataValidateDetails: Record<string, MetadataValidateDetailsItem>;
+}
+
+export interface InitialState {
+  sampleNameColumn: string;
+  headers: string[];
+  metadata: MetadataItem[];
+  metadataValidateDetails: Record<string, MetadataValidateDetailsItem>;
+  metadataSaveDetails: Record<string, MetadataSaveDetailsItem>;
+}
+
+const initialState: InitialState = {
+  sampleNameColumn: "",
+  headers: [],
+  metadata: [],
+  metadataValidateDetails: {},
+  metadataSaveDetails: {},
+};
+
+/*
+Redux async thunk for saving the metadata to samples.
+For more information on redux async thunks see: https://redux-toolkit.js.org/api/createAsyncThunk
+ */
+export const saveMetadata = createAsyncThunk<
+  SaveMetadataResponse,
+  { projectId: string; selectedMetadataKeys: string[] },
+  { dispatch: ImportDispatch; state: ImportState }
+>(
+  `importReducer/saveMetadata`,
+  async ({ projectId, selectedMetadataKeys }, { dispatch, getState }) => {
+    const state: ImportState = getState();
+    const sampleNameColumn: string = state.importReducer.sampleNameColumn;
```

Comment:
Just use JavaScript object destructuring for these lines since they all pull attributes from the same object.  We have an eslint rule to check for this.

Reply:
Please see a50d6d745b0f1bc4792cda3705ae915fbfe0ec5e.